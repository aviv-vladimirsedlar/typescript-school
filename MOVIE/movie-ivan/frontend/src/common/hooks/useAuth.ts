import { useSelector } from 'react-redux';

import { RootState } from '../../config/store';

export const useAuth = () => {
  return useSelector((state: RootState) => state.auth.isAuthenticated);
};
