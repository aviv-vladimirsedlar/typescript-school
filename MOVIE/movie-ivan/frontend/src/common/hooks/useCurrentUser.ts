import { useState } from 'react';
import { useQuery } from 'react-query';
import { useDispatch } from 'react-redux';

import axiosInstance from '../../config/api';
import { loginSuccess } from '../../config/slices/auth.slice';

const getCurrentUser = async () => {
  const response = await axiosInstance.get('/users/me');
  return response.data;
};

export const useCurrentUser = () => {
  const [currentUser, setCurrentUser] = useState(null);
  const dispatch = useDispatch();

  const { isLoading } = useQuery('currentUser', getCurrentUser, {
    retry: false,
    onSuccess: (res) => {
      setCurrentUser(res);
      dispatch(loginSuccess({ user: res }));
    },
    onError: (error) => {
      console.error('Login error:', error);
    },
  });

  return { currentUser, isLoading };
};
