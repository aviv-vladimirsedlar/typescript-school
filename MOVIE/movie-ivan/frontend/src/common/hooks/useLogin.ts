import { useMutation } from 'react-query';
import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import axiosInstance from '../../config/api';
import { loginSuccess } from '../../config/slices/auth.slice';

interface LoginData {
  email: string;
  password: string;
}

const loginRequest = async (data: LoginData) => {
  const response = await axiosInstance.post('/users/login', data);
  return response.data;
};

export const useLogin = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  return useMutation(loginRequest, {
    onSuccess: (data) => {
      dispatch(loginSuccess({ user: data.user }));
      navigate('/');
    },
  });
};
