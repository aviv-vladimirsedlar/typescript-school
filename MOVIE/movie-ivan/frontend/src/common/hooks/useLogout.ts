import { useMutation } from 'react-query';
import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import axiosInstance from '../../config/api';
import { logoutSuccess } from '../../config/slices/auth.slice';

const logoutRequest = async () => {
  const response = await axiosInstance.delete('/users/logout');
  return response.data;
};

export const useLogout = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  return useMutation(logoutRequest, {
    onSuccess: () => {
      dispatch(logoutSuccess());
      setTimeout(() => {
        navigate('/auth/login');
      }, 1000);
    },
    onError: (error) => {
      console.error('Login error:', error);
    },
  });
};
