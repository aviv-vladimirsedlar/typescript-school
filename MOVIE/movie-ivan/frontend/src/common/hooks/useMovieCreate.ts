import { useMutation } from 'react-query';

import axiosInstance from '../../config/api';
import { Movie } from '../types/movie.types';

type MovieData = Pick<Movie, 'title' | 'description' | 'year' | 'duration'>;

const createMovie = async (data: { movieId: string; data: MovieData }) => {
  const response = await axiosInstance.post(`/movies/create`, data.data);
  return response.data;
};

export const useMovieCreate = () => {
  return useMutation(createMovie, {
    onError: (error) => {
      console.error('Movie create error:', error);
    },
  });
};
