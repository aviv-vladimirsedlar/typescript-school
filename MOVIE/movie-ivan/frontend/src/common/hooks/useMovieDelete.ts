import { useMutation } from 'react-query';

import axiosInstance from '../../config/api';

const deleteMovie = async (movieId: string) => {
  const response = await axiosInstance.delete(`/movies/${movieId}`);
  return response.data;
};

export const useMovieDelete = ({ onSuccess }: { onSuccess: () => void }) => {
  return useMutation(deleteMovie, {
    onSuccess,
    onError: (error) => {
      console.error('Delete error:', error);
    },
  });
};
