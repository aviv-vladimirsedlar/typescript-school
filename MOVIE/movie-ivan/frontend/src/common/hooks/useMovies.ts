import { useQuery } from 'react-query';
import { useDispatch, useSelector } from 'react-redux';

import axiosInstance from '../../config/api';
import { updateMovieList } from '../../config/slices/movie.slice';
import { RootState } from '../../config/store';

const getMovies = async () => {
  const response = await axiosInstance.get('/movies?limit=100');
  return response.data;
};

export const useMovies = () => {
  const moviesData = useSelector((state: RootState) => {
    return state.movie;
  });

  const dispatch = useDispatch();

  const { isLoading, refetch } = useQuery('getMovies?limit=100', getMovies, {
    retry: false,
    onSuccess: (data) => {
      dispatch(updateMovieList(data));
    },
    onError: (error) => {
      console.error('Movie list error:', error);
    },
  });

  return { data: moviesData.data, meta: moviesData.meta, isLoading, refetch };
};
