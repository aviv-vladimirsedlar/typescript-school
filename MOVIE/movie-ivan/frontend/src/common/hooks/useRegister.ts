import { useMutation } from 'react-query';
import { useDispatch } from 'react-redux';

import axiosInstance from '../../config/api';
import { loginSuccess } from '../../config/slices/auth.slice';

interface RegisterData {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
}

const registerRequest = async (data: RegisterData) => {
  const response = await axiosInstance.post('/users/register', data);
  return response.data;
};

export const useRegister = () => {
  const dispatch = useDispatch();

  return useMutation(registerRequest, {
    onSuccess: (data) => {
      dispatch(loginSuccess({ user: data.user }));
    },
  });
};
