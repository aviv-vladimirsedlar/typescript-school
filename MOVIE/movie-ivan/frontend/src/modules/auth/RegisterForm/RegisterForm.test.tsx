import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import React from 'react';

import '@testing-library/jest-dom';
import { useRegister } from '../../../common/hooks/useRegister';

import { RegisterForm } from '.';

jest.mock('../../../common/hooks/useRegister');

const mockMutate = jest.fn();
const mockUseRegister = useRegister as jest.MockedFunction<typeof useRegister>;

describe('RegisterForm', () => {
  beforeEach(() => {
    // Reset mocks before each test
    jest.clearAllMocks();
    mockUseRegister.mockReturnValue({
      mutate: mockMutate,
      isLoading: false,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any);
  });

  it('should match snapshot when initially rendered', () => {
    const { asFragment } = render(<RegisterForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the form and handle successful registration', async () => {
    mockMutate.mockImplementation((data, { onSuccess }) => onSuccess && onSuccess());
    render(<RegisterForm />);
    await act(async () => {
      fireEvent.change(screen.getByTestId('first-name-input'), { target: { value: 'John' } });
      fireEvent.change(screen.getByTestId('last-name-input'), { target: { value: 'Doe' } });
      fireEvent.change(screen.getByTestId('email-input'), { target: { value: 'john.doe@example.com' } });
      fireEvent.change(screen.getByTestId('password-input'), { target: { value: 'Password123' } });
      fireEvent.change(screen.getByTestId('confirm-password-input'), { target: { value: 'Password123' } });
      fireEvent.click(screen.getByTestId('btn-register'));
    });
    await waitFor(() => expect(mockMutate).toHaveBeenCalled());
    expect(screen.queryByText(/An unknown error occurred/i)).not.toBeInTheDocument();
  });

  it('should show error message on failed registration', async () => {
    mockMutate.mockImplementation(
      (data, { onError }) => onError && onError({ response: { data: { message: 'Registration failed' } } }),
    );
    render(<RegisterForm />);
    await act(async () => {
      fireEvent.change(screen.getByTestId('first-name-input'), { target: { value: 'John' } });
      fireEvent.change(screen.getByTestId('last-name-input'), { target: { value: 'Doe' } });
      fireEvent.change(screen.getByTestId('email-input'), { target: { value: 'john.doe@example.com' } });
      fireEvent.change(screen.getByTestId('password-input'), { target: { value: 'Password123' } });
      fireEvent.change(screen.getByTestId('confirm-password-input'), { target: { value: 'Password123' } });
      fireEvent.click(screen.getByTestId('btn-register'));
    });
    await waitFor(() => expect(screen.getByText(/Registration failed/i)).toBeInTheDocument());
  });

  it('should validate required fields and show validation errors', async () => {
    render(<RegisterForm />);
    await act(async () => {
      fireEvent.click(screen.getByTestId('btn-register'));
    });
    await waitFor(() => {
      expect(screen.getByTestId('first-name-input')).toHaveClass(/border-red-/i);
      expect(screen.getByTestId('last-name-input')).toHaveClass(/border-red-/i);
      expect(screen.getByTestId('email-input')).toHaveClass(/border-red-/i);
      expect(screen.getByTestId('password-input')).toHaveClass(/border-red-/i);
      expect(screen.getByTestId('confirm-password-input')).toHaveClass(/border-red-/i);
    });
  });
});
