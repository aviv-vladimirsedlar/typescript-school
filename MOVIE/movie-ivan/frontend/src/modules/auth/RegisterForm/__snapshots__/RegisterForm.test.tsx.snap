// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm should match snapshot when initially rendered 1`] = `
<DocumentFragment>
  <form
    class="mx-auto w-full"
  >
    <div
      class="grid grid-cols-2 gap-4"
    >
      <div
        class="mb-5"
      >
        <label
          aria-labelledby="firstName"
          class="mb-2 block text-sm font-medium text-gray-900 dark:text-white"
          for="firstName"
        >
          First name 
          <span>
            *
          </span>
        </label>
        <input
          class="block w-full rounded-lg bg-gray-50/50 p-4 text-sm text-gray-900 border border-gray-200"
          data-testid="first-name-input"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-5"
      >
        <label
          aria-labelledby="lastName"
          class="mb-2 block text-sm font-medium text-gray-900 dark:text-white"
          for="lastName"
        >
          Last name 
          <span>
            *
          </span>
        </label>
        <input
          class="block w-full rounded-lg bg-gray-50/50 p-4 text-sm text-gray-900 border border-gray-200"
          data-testid="last-name-input"
          required=""
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="mb-5"
    >
      <label
        aria-labelledby="email"
        class="mb-2 block text-sm font-medium text-gray-900 dark:text-white"
        for="email"
      >
        Your email 
        <span>
          *
        </span>
      </label>
      <input
        class="block w-full rounded-lg bg-gray-50/50 p-4 text-sm text-gray-900 border border-gray-200"
        data-testid="email-input"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-5"
    >
      <label
        aria-labelledby="password"
        class="mb-2 block text-sm font-medium text-gray-900 dark:text-white"
        for="password"
      >
        Password 
        <span>
          *
        </span>
      </label>
      <input
        class="block w-full rounded-lg bg-gray-50/50 p-4 text-sm text-gray-900 border border-gray-200"
        data-testid="password-input"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-5"
    >
      <label
        aria-labelledby="passwordConfirm"
        class="mb-2 block text-sm font-medium text-gray-900 dark:text-white"
        for="passwordConfirm"
      >
        Confirm password 
        <span>
          *
        </span>
      </label>
      <input
        class="block w-full rounded-lg bg-gray-50/50 p-4 text-sm text-gray-900 border border-gray-200"
        data-testid="confirm-password-input"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="rounded-lg bg-blue-500 p-2 px-6 font-semibold w-full text-white"
      data-testid="btn-register"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
